cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/actor/ProgressiveSolver.scala

```diff
@@ -0,0 +1,180 @@
+        }
+        val schema = info.schema.asInstanceOf[Schema]
+        val bound = query.getTimeInterval(schema.timeField).getOrElse(new TInterval(info.dataInterval.getStart, DateTime.now))
+        val merger = planner.calculateMergeFunc(query, info.schema)
+        val queryWOTime = query.copy(filter = query.filter.filterNot(_.field == schema.timeField))
+        MiniQuery(queryWOTime, info, bound, merger)
+      }
+      val min = queryInfos.map(_.queryBound.getStartMillis).min
+      val max = queryInfos.map(_.queryBound.getEndMillis).max
+      val boundary = new TInterval(min, max)
+
+      val initialDuration = config.FirstQueryTimeGap
+      val interval = calculateFirst(boundary, initialDuration)
+      val queryGroup = QueryGroup(ts, queryInfos, request.postTransform)
+      val initResult = Seq.fill(queryInfos.size)(JsArray())
+      issueQueryGroup(interval, queryGroup)
+      val drumEstimator = new Drum(boundary.toDuration.getStandardHours.toInt, alpha = 1, initialDuration.toHours.toInt)
+      context.become(askSlice(reporter, request.resultSizeLimitOpt, request.intervalMS, request.intervalMS, interval, drumEstimator, Int.MaxValue, boundary, queryGroup, initResult, issuedTimestamp = DateTime.now), discardOld = true)
+    case _: MiniQueryResult =>
+      // do nothing
+      log.debug(s"receive: obsolete query result")
+    case Cancel =>
+      // do nothing
+      log.debug(s"receive: cancel")
+  }
+
+  private def askSlice(reporter: ActorRef,
+                       resultSizeLimitOpt: Option[Int],
+                       paceMS: Long,
+                       timeLimitMS: Long,
+                       curInterval: TInterval,
+                       estimator: Drum,
+                       curEstimateMS: Long,
+                       boundary: TInterval,
+                       queryGroup: QueryGroup,
+                       accumulateResults: Seq[JsArray],
+                       issuedTimestamp: DateTime): Receive = {
+    case result: MiniQueryResult if result.key == ts =>
+      val mergedResults = queryGroup.queries.zipWithIndex.map {
+        case (q, idx) =>
+          q.merger(Seq(accumulateResults(idx), result.jsons(idx)))
+      }
+
+      val timeSpend = DateTime.now.getMillis - issuedTimestamp.getMillis
+      val diff = Math.max(0, timeLimitMS - timeSpend)
+      val nextLimit = paceMS + diff
+
+      val (nextInterval, nextEstimateMS) = calculateNext(estimator, nextLimit, curInterval, curEstimateMS, timeSpend, boundary)
+      if (nextInterval.toDurationMillis == 0 || hasEnoughResults(mergedResults, resultSizeLimitOpt)) { //finished slicing
+
+        val limitResultOpt = resultSizeLimitOpt.map(limit => Seq(JsArray(mergedResults.head.value.take(limit))))
+        val returnedResult = limitResultOpt.getOrElse(mergedResults)
+        reporter ! Reporter.PartialResult(curInterval.getStartMillis, boundary.getEndMillis, 1.0, queryGroup.postTransform.transform(JsArray(returnedResult)))
```

Comment:
The `progress` 1.0 in `ParsialResult` is configurable? Why don't we use `boundary.getEndMillis - boundary.getStartMillis`?

Reply:
Where is this `progress` meant to use (in the future)?

Reply:
1.0 means 100%. This is the last response whose progress indicates the process is over. 
In other cases, it gives users a progress value, e.g. 0.75 that means 75% of the work has been done.

Reply:
Got it.